File: day17/part2.py
import argparse
import collections
from typing import Callable
from typing import Dict
from typing import List
from typing import Tuple

from support import timing


def run(
        prog: Dict[int, int],
        in_fn: Callable[[], int],
        out_fn: Callable[[int], None],
) -> None:
    def parameter(instr: int, n: int) -> int:
        mode = instr // (10 ** (n + 1)) % 10
        if mode == 0:
            return prog[prog[pc + n]]
        elif mode == 1:
            return prog[pc + n]
        elif mode == 2:
            return prog[rb + prog[pc + n]]
        else:
            raise NotImplementedError(mode)

    def store(instr: int, n: int) -> int:
        mode = instr // (10 ** (n + 1)) % 10
        if mode == 0:
            return prog[pc + n]
        elif mode == 2:
            return rb + prog[pc + n]
        else:
            raise NotImplementedError(mode)

    rb = 0
    pc = 0
    while pc < len(prog):
        instr = prog[pc]
        opc = instr % 100
        if opc == 99:
            return
        elif opc == 1:
            prog[store(instr, 3)] = parameter(instr, 1) + parameter(instr, 2)
            pc += 4
        elif opc == 2:
            prog[store(instr, 3)] = parameter(instr, 1) * parameter(instr, 2)
            pc += 4
        elif opc == 3:
            prog[store(instr, 1)] = in_fn()
            pc += 2
        elif opc == 4:
            out_fn(parameter(instr, 1))
            pc += 2
        elif opc == 5:
            if parameter(instr, 1):
                pc = parameter(instr, 2)
            else:
                pc += 3
        elif opc == 6:
            if not parameter(instr, 1):
                pc = parameter(instr, 2)
            else:
                pc += 3
        elif opc == 7:
            if parameter(instr, 1) < parameter(instr, 2):
                prog[store(instr, 3)] = 1
            else:
                prog[store(instr, 3)] = 0
            pc += 4
        elif opc == 8:
            if parameter(instr, 1) == parameter(instr, 2):
                prog[store(instr, 3)] = 1
            else:
                prog[store(instr, 3)] = 0
            pc += 4
        elif opc == 9:
            rb += parameter(instr, 1)
            pc += 2
        else:
            raise AssertionError(f'unreachable? {prog} {pc}')
    raise AssertionError(f'unreachable? {prog} {pc}')


def print_grid(grid: Dict[Tuple[int, int], str]) -> None:
    min_x = min(x for x, _ in grid)
    max_x = max(x for x, _ in grid)
    min_y = min(y for _, y in grid)
    max_y = max(y for _, y in grid)
    for y in range(min_y, max_y + 1):
        for x in range(min_x, max_x + 1):
            print(grid[x, y], end='')
        print()
    print('=' * 79)


DIRECTIONS = [(0, -1), (1, 0), (0, 1), (-1, 0)]


def add_coord(p1: Tuple[int, int], p2: Tuple[int, int]) -> Tuple[int, int]:
    return p1[0] + p2[0], p1[1] + p2[1]


def get_path(
        grid: Dict[Tuple[int, int], str],
        pos: Tuple[int, int],
) -> List[str]:
    assert grid[pos] == '^'
    direction = DIRECTIONS[0]
    ret: List[str] = []
    n = 0

    while '#' in grid.values():
        pos_idx = DIRECTIONS.index(direction)
        right_dir = DIRECTIONS[(pos_idx + 1) % 4]
        left_dir = DIRECTIONS[(pos_idx - 1) % 4]
        next_straight = add_coord(pos, direction)
        next_right = add_coord(pos, right_dir)
        next_left = add_coord(pos, left_dir)
        if grid[next_straight] in {'#', 'X'}:
            n += 1
            pos = next_straight
            grid[pos] = 'X'
        elif grid[next_right] == '#':
            ret.extend((str(n), 'R'))
            n = 0
            direction = right_dir
        else:
            assert grid[next_left] == '#'
            ret.extend((str(n), 'L'))
            n = 0
            direction = left_dir

    ret.append(str(n))
    return ret[1:]


INPUT = '''\
#######...#####
#.....#...#...#
#.....#...#...#
......#...#...#
......#...###.#
......#.....#.#
^########...#.#
......#.#...#.#
......#########
........#...#..
....#########..
....#...#......
....#...#......
....#...#......
....#####......
'''


def test_get_path() -> None:
    bot_x = 0
    bot_y = 0
    grid: Dict[Tuple[int, int], str] = collections.defaultdict(str)
    for y, row in enumerate(INPUT.splitlines()):
        for x, s in enumerate(row):
            grid[x, y] = s
            if s == '^':
                bot_x = x
                bot_y = y
    assert get_path(grid, (bot_x, bot_y)) == [
        'R', '8', 'R', '8', 'R', '4', 'R', '4', 'R', '8', 'L', '6', 'L', '2',
        'R', '4', 'R', '4', 'R', '8', 'R', '8', 'R', '8', 'L', '6', 'L', '2',
    ]


def possible_strings(
        s: str,
        s_idx: int,
        prev: Tuple[str, ...],
) -> List[Tuple[str, int]]:
    e_idx = s_idx

    while True:
        for prev_s in prev:
            if s.startswith(prev_s, s_idx):
                s_idx = e_idx = s_idx + len(prev_s) + 1
                break
        else:
            break

    ret = []
    while True:
        e_idx = s.find(',', e_idx + 1)
        e_idx = s.find(',', e_idx + 1)
        if s.count(s[s_idx:e_idx]) == 1 or e_idx - s_idx > 20:
            break
        else:
            ret.append((s[s_idx:e_idx], e_idx + 1))
    return ret


def get_prog(prog: List[str]) -> str:
    s = ','.join(prog)

    possible = []
    for s1, s1_e in possible_strings(s, 0, ()):
        for s2, s2_e in possible_strings(s, s1_e, (s1,)):
            for s3, s3_e in possible_strings(s, s2_e, (s1, s2)):
                possible.append((s1, s2, s3))

    for ret_strings in possible:
        main = []
        s_try = s
        while s_try:
            for i, routine in enumerate(ret_strings):
                if s_try.startswith(routine):
                    s_try = s_try[len(routine) + 1:]
                    main.append(chr(ord('A') + i))
                    break
            else:
                break
        else:
            if len(','.join(main)) < 20:
                return '\n'.join((','.join(main), *ret_strings, 'n', ''))
    else:
        raise AssertionError('unreachable')


def test_get_prog() -> None:
    input_actions = [
        'R', '8', 'R', '8', 'R', '4', 'R', '4', 'R', '8', 'L', '6', 'L', '2',
        'R', '4', 'R', '4', 'R', '8', 'R', '8', 'R', '8', 'L', '6', 'L', '2',
    ]
    # given answer as this, but ours also works :shrug:
    # A,B,C,B,A,C
    # R,8,R,8
    # R,4,R,4,R,8
    # L,6,L,2
    # n
    assert get_prog(input_actions) == '''\
A,A,B,A,C,B,A,A,A,C
R,8
R,4,R,4
L,6,L,2
n
'''


def test_get_prog2() -> None:
    input_actions = [
        'L', '4', 'R', '8', 'L', '6', 'L', '10', 'L', '6', 'R', '8',
        'R', '10', 'L', '6', 'L', '6', 'L', '4', 'R', '8', 'L', '6',
        'L', '10', 'L', '6', 'R', '8', 'R', '10', 'L', '6', 'L', '6',
        'L', '4', 'L', '4', 'L', '10', 'L', '4', 'L', '4', 'L', '10',
        'L', '6', 'R', '8', 'R', '10', 'L', '6', 'L', '6', 'L', '4',
        'R', '8', 'L', '6', 'L', '10', 'L', '6', 'R', '8', 'R', '10',
        'L', '6', 'L', '6', 'L', '4', 'L', '4', 'L', '10',
    ]
    assert get_prog(input_actions) == '''\
A,B,A,B,C,C,B,A,B,C
L,4,R,8,L,6,L,10
L,6,R,8,R,10,L,6,L,6
L,4,L,4,L,10
n
'''


def compute(s: str) -> int:
    prog = [int(part) for part in s.strip().split(',')]
    prog_d = collections.defaultdict(int, enumerate(prog))
    prog_d[0] = 2

    grid: Dict[Tuple[int, int], str] = collections.defaultdict(str)

    answer = 0
    y = 0
    x = 0
    bot_x = 0
    bot_y = 0

    computed = False
    machine_prog = iter('')

    def in_fn() -> int:
        nonlocal computed
        nonlocal machine_prog

        if not computed:
            assert bot_x != 0, bot_x
            assert bot_y != 0, bot_y
            assert grid[bot_x, bot_y] == '^', grid[bot_x, bot_y]
            path = get_path(grid, (bot_x, bot_y))
            machine_prog = iter(get_prog(path))
            computed = True

        return ord(next(machine_prog))

    def out_fn(n: int) -> None:
        nonlocal answer
        nonlocal x
        nonlocal y
        nonlocal bot_x
        nonlocal bot_y

        if n < 255:
            s = chr(n)
            if s == '\n':
                x = 0
                y += 1
            else:
                if s == '^':
                    bot_x = x
                    bot_y = y
                grid[x, y] = s
                x += 1
        else:
            answer = n

    run(prog_d, in_fn, out_fn)

    return answer


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('data_file')
    args = parser.parse_args()

    with open(args.data_file) as f, timing():
        print(compute(f.read()))

    return 0


if __name__ == '__main__':
    exit(main())

File: day17/part2.php
<?php

if (count($argv) !== 2) {
    echo 'usage: ' . $argv[0] . " FILENAME\n";
}

$prog = json_decode('[' . file_get_contents($argv[1]) . ']', true);
$prog[0] = 2;

$rb = 0;
$pc = 0;

$x = 0;
$y = 0;
$bot_x = 0;
$bot_y = 0;
$grid = array(array());

$run_once = FALSE;
$path = array();
$out_prog = '';

function add_coord($p1, $p2) {
    return [$p1[0] + $p2[0], $p1[1] + $p2[1]];
}

function print_grid($grid) {
    foreach ($grid as $y_coord => $row) {
        foreach ($row as $x_coord => $s) {
            echo $s;
        }
        echo "\n";
    }
}

function calculate_path($grid, $pos) {
    assert($grid[$pos[1]][$pos[0]] == '^');
    $DIRECTIONS = [[0, -1], [1, 0], [0, 1], [-1, 0]];
    $direction = $DIRECTIONS[0];
    $ret = array();
    $n = 0;

    while (TRUE) {
        $pos_idx = array_search($direction, $DIRECTIONS);
        $right_dir = $DIRECTIONS[($pos_idx + 1) % 4];
        $left_dir = $DIRECTIONS[($pos_idx + 3) % 4];
        $next_up = add_coord($pos, $direction);
        $next_right = add_coord($pos, $right_dir);
        $next_left = add_coord($pos, $left_dir);

        if (in_array(($grid[$next_up[1]][$next_up[0]] ?? ''), ['#', 'X'])) {
            $n += 1;
            $pos = $next_up;
            $grid[$pos[1]][$pos[0]] = 'X';
        } else if (($grid[$next_right[1]][$next_right[0]] ?? '') === '#') {
            array_push($ret, strval($n), 'R');
            $n = 0;
            $direction = $right_dir;
        } else if (($grid[$next_left[1]][$next_left[0]] ?? '') === '#') {
            array_push($ret, strval($n), 'L');
            $n = 0;
            $direction = $left_dir;
        } else {
            array_push($ret, strval($n));
            return array_slice($ret, 1);
        }
    }
};

function calculate_prog($path) {

}

function param($instr, $n, $prog, $rb, $pc) {
    $mode = intdiv($instr, pow(10, ($n + 1))) % 10;
    switch ($mode) {
        case 0:
            return $prog[$prog[$pc + $n] ?? 0] ?? 0;
        case 1:
            return $prog[$pc + $n] ?? 0;
        case 2:
            return $prog[$rb + $prog[$pc + $n] ?? 0] ?? 0;
        default:
            assert("unreachable");
    }
};

function store($instr, $n, $prog, $rb, $pc) {
    $mode = intdiv($instr, pow(10, ($n + 1))) % 10;
    switch ($mode) {
        case 0:
            return $prog[$pc + $n] ?? 0;
        case 2:
            return $rb + $prog[$pc + $n] ?? 0;
        default:
            assert("unreachable");
    }
}

while (TRUE) {
    $instr = $prog[$pc];
    $opc = $instr % 100;
    switch ($opc) {
        case 99:
            break 2;
        case 1:
            $prog[store($instr, 3, $prog, $rb, $pc)] = param($instr, 1, $prog, $rb, $pc) + param($instr, 2, $prog, $rb, $pc);
            $pc += 4;
            break;
        case 2:
            $prog[store($instr, 3, $prog, $rb, $pc)] = param($instr, 1, $prog, $rb, $pc) * param($instr, 2, $prog, $rb, $pc);
            $pc += 4;
            break;
        case 3:
            if (!$run_once) {
                while (empty($grid[count($grid) - 1])) {
                    unset($grid[count($grid) - 1]);
                }
                $path = calculate_path($grid, [$bot_x, $bot_y]);
                $out_prog = calculate_prog($path);
                echo $out_prog . "\n";
                exit(1);
                $run_once = TRUE;
            }
            $prog[store($instr, 1, $prog, $rb, $pc)] = ord($out_prog[$idx]);
            $idx += 1;
            $pc += 2;
            break;
        case 4:
            $s = chr(param($instr, 1, $prog, $rb, $pc));
            if ($s === "\n") {
                $x = 0;
                $y += 1;
                $grid[$y] = array();
            } else if (in_array($s, ['.', '#', '^'])) {
                if ($s === '^') {
                    $bot_x = $x;
                    $bot_y = $y;
                }
                $grid[$y][$x] = $s;
                $x += 1;
            }
            echo $s;
            $pc += 2;
            break;
        case 5:
            if (param($instr, 1, $prog, $rb, $pc) !== 0) {
                $pc = param($instr, 2, $prog, $rb, $pc);
            } else {
                $pc += 3;
            }
            break;
        case 6:
            if (param($instr, 1, $prog, $rb, $pc) === 0) {
                $pc = param($instr, 2, $prog, $rb, $pc);
            } else {
                $pc += 3;
            }
            break;
        case 7:
            if (param($instr, 1, $prog, $rb, $pc) < param($instr, 2, $prog, $rb, $pc)) {
                $prog[store($instr, 3, $prog, $rb, $pc)] = 1;
            } else {
                $prog[store($instr, 3, $prog, $rb, $pc)] = 0;
            }
            $pc += 4;
            break;
        case 8:
            if (param($instr, 1, $prog, $rb, $pc) === param($instr, 2, $prog, $rb, $pc)) {
                $prog[store($instr, 3, $prog, $rb, $pc)] = 1;
            } else {
                $prog[store($instr, 3, $prog, $rb, $pc)] = 0;
            }
            $pc += 4;
            break;
        case 9:
            $rb += param($instr, 1, $prog, $rb, $pc);
            $pc += 2;
        default:
            assert('unreachable');
    }
}
